### PR TITLE
Fix/tabs discard dead tab

### LINKED ISSUE #487: Implement chrome.tabs.onReplaced.addEventListener
When a tab becomes `discarded` from a previously-loaded state, its tabId is changed permanently. The previous Tab/WebContents is destroyed, and a new Tab/WebContents is created. The same happens for Chrome as for muon. However, in Chrome, [an event (`tabs.onReplaced`)](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/tabs/onReplaced) is exposed which provides the translation between the oldTabId and the newTabId.

This is important for consuming apps in order to present new incoming Tab/WebContents objects as 'new' by animating in, and not performing that animation for new Tab/WebContents objects which are previously known to the app. Since there is no other property to represent the Tab object uniquely, there is no other way to differentiate a 'new' tab versus a 'replaced' tab.

The only way to do this currently is to always have the tab attached to a `<webview>` object in a renderer window, and handle the `tab-id-changed` event inside the renderer window. This prevents being able to have only the active tab attached to a renderer window at a time, which has performance implications.

image

### LINKED ISSUE #496: Tab is destroyed when detached if it has previously been discarded
Even if the tab has been reloaded since it was discarded, detach will result in a destroyed tab.

STR:

1. Create a tab, load contents, and make a different tab active
1. `tab.discard()` on the original tab
  - the tab ID will be changed
2. Make the tab active so that it reloads it's content
   - the tab ID will stay the same
3. `tab.detach()` on the new web contents
4. The tab will be destroyed

Event log:
```
tab[33].detach(): { windowId: 1, guestInstance: 6 }
 - [37] on-web-contents-created
 - [33] will-detach
 - [33] tab-replaced-at { newTabId 37, newTab: { windowId: -1, guestInstance: 7 } }
 - [33] will-destroy
 - [33] destroyed
 - [33] render-view-deleted
 - [37] will-attach { windowId: 1 }
 - [33] <webview>.tab-id-changed { tabId: 37 }
 - [37] will-destroy
 - [37] destroyed
new BrowserWindow()-->entry.js-->addFrame({tabId: 33})
// tab not found!
```

Expected event log (real case of a tab that has *never* been discarded:
```
tab[13].detach() current state - { windowId: 1, guestInstance: 2 }
 - [37] web-contents-created
 - [13] will-detach
 - [13] tab-replaced-at { newTabId: 37, newTab: { windowId: -1, guestInstance: 7 }
 - [13] <webview>.tab-id-changed { tabId: 37 }
 - [13] did-detach
 - [13] chrome-tabs-updated { index: -1 }
 - [37] will-attach { windowId: 1 }
 - [37] will-destroy
 - [37] destroyed
 - [37] render-view-deleted (x2)
 - [13] <webview>.tab-id-changed { tabId: 37 }
new BrowserWindow()-->entry.js-->addFrame({tabId: 13})
 - [13] will-attach
 - [13] tab-inserted-at
 - [13] set-active
 - [13] did-attach { windowId: 4}
 - [13] render-view-ready
```

### LINKED ISSUE #505: Detaching a discarded tab and attaching to a new window causes a duplicate navigation entry to be added
...and scroll position / input entry state to be lost.
Discovered whilst testing #504 

## STR
Make sure you have #504 and https://github.com/brave/browser-laptop/pull/13120
Login to github.com
Visit http://github.com/brave/browser-laptop/issues/new
Enter "TEST" at the top of the issue description
Switch to another tab
Right-click the github issue tab and select 'Discard'
Right-click the github issue tab (be careful not to trigger a preview) and choose 'Detach'

### Actual
The detached, and now reloaded tab, does not contain the word "TEST" at the top of the issue description.
The tab now has a additional 'back' navigation entry to the same Url.

### Expected
The draft description is preserved.
This works when discarding a tab and then making it active, causing a reload. The draft content is preserved. Adding the action of a detach and then attach to a new window is causing some data to be lost.

Tested that session storage, local storage and session cookies are preserved ✅

### PR DESCRIPTION
**Pull request by bridiver**
Fix #487 
Fix #496 
Fix #505

### CODE REVIEW

**Comment by petemill** [comment]
Added that this fixes #487.

**Comment by petemill** [comment]
@bsclifton I approved but new commits were pushed marking it 'stale'. FWIW all 3 issues are resolved as far as I'm concerned, and I've found no new issues.

File: brave/browser/guest_view/tab_view/tab_view_guest.cc [hunk 1]
```diff
@@ -318,6 +318,11 @@ void TabViewGuest::ApplyAttributes(const base::DictionaryValue& params) {
 
       if (attached() &&
           web_contents()->GetController().IsInitialNavigation()) {
+        // don't reload if we're already loading
+        if (web_contents()->GetController().GetPendingEntry() &&
+            web_contents()->GetController().GetPendingEntry()->GetURL() == src_) {
```
Comment (darkdh) [inline brave/browser/guest_view/tab_view/tab_view_guest.cc]: lint: >80 chars
